### PR TITLE
Implement _so_ many things, upgrade all deps.

### PR DESCRIPTION
**Pull request by illegalprime**
This code change is ~~**not yet complete, and passes no tests**~~ **complete, needs docs**.

Summary of changes:
__a lot, so coming soon__

### CODE REVIEW

File: src/server/upgrade/hyper.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	fn into_ws(self) -> Result<WsUpgrade<Self::Stream>, Self::Error> {
+		if let Err(e) = validate(&self.method, &self.version, &self.headers) {
+			return Err((self, e));
+		}
+
+		let stream = if let Some(http) = self.downcast_ref::<HttpStream>() {
+			http.0.try_clone().ok().map(Box::new)
+		} else if let Some(ssl) = self.downcast_ref::<SslStream<HttpStream>>() {
+			unimplemented!();
+		} else if let Some(ssl) = self.downcast_ref::<HttpsStream<SslStream<HttpStream>>>() {
```
Comment (illegalprime) [inline src/server/upgrade/hyper.rs]: There must be a better way!

**Comment by cyderize** [comment]
@illegalprime Thanks for this!

These changes seem good to me, I can't really see anything that I think should be done differently.

Awesome work!

File: src/server/upgrade/hyper.rs (hunk 1 above)
Comment (cyderize) [inline src/server/upgrade/hyper.rs]: Hmm, yeah this doesn't seem ideal really - I can't think of a better way right now though, haha

**Comment by djc** [comment]
@cyderize so what needs to happen to land this? Servo would really like to use this.

**Comment by DenisKolodin** [comment]
I quite like this changes.
@cyderize @illegalprime What is the fate of this PR?

**Comment by beatgammit** [comment]
Any update on this?

**Comment by vi** [comment]
@beatgammit, Note: forked crate `websocket` into `websocket-vi`. Are you interested in rust-websocket continuation? Do you want me to try fixing and integrating this pull request?

**Comment by illegalprime** [comment]
@vi @beatgammit yeah I've had quite a year of failures on my part :cry: I am refocusing and would love more help! I guess we should make a list of TODOs like in #27

**Comment by illegalprime** [comment]
@cyderize I'm pretty close to finishing this up, basically everything from #58 is implemented except for `permessage-deflate` and mio/tokio. All that's left is docs and tests.

Since this change is rather large I wanted to give you a heads up if you wanted to review all the changes.

**Pull request merged**